### ISSUE TITLE
field components should consistently support isReadOnly and isRequired render-props+data-attributes

### ISSUE THREAD

**Issue by stefanprobst** [adobe/react-spectrum#6151]
### Provide a general summary of the feature here

this is continuing https://github.com/adobe/react-spectrum/issues/5930#issuecomment-1988997419

field components (i.e. those which can be used in a form) should consistently support disabled/invalid/readonly/required props, render-props, and data attributes.

i have tried to list the current support here (although not every case may make sense to support). also, i have not yet created a separate table for data-attributes, but it seems that these match render-props support:

## props

| RAC            | isDisabled | isInvalid | isReadOnly | isRequired |
| -------------- | ---------- | --------- | ---------- | ---------- |
| CheckboxGroup  | ✅         | ✅        | ✅         | ✅         |
| ComboBox       | ✅         | ✅        | ✅         | ✅         |
| Date/TimeField | ✅         | ✅        | ✅         | ✅         |
| NumberField    | ✅         | ✅        | ✅         | ✅         |
| RadioGroup     | ✅         | ✅        | ✅         | ✅         |
| SearchField    | ✅         | ✅        | ✅         | ✅         |
| Select         | ✅         | ✅        | ❌         | ✅         |
| Slider         | ✅         | ❌        | ❌         | ❌         |
| Switch         | ✅         | ❌        | ✅         | ❌         |
| TextField      | ✅         | ✅        | ✅         | ✅         |

## render props

| RAC            | isDisabled | isInvalid | isReadOnly | isRequired |
| -------------- | ---------- | --------- | ---------- | ---------- |
| CheckboxGroup  | ✅         | ✅        | ✅         | ✅         |
| ComboBox       | ✅         | ✅        | ❌         | ✅         |
| Date/TimeField | ✅         | ✅        | ✅         | ✅         |
| NumberField    | ✅         | ✅        | ❌         | ✅         |
| RadioGroup     | ✅         | ✅        | ✅         | ✅         |
| SearchField    | ✅         | ✅        | ✅         | ✅         |
| Select         | ✅         | ✅        | ❌         | ✅         |
| Slider         | ✅         | ❌        | ❌         | ❌         |
| Switch         | ✅         | ❌        | ✅         | ❌         |
| TextField      | ✅         | ✅        | ✅         | ✅         |

### 🤔 Expected Behavior?

see above

### 😯 Current Behavior

see above

### 🔦 Context

see above

**Comment by snowystinger** [comment]
Thanks for the issue and the details
Related https://github.com/adobe/react-spectrum/issues/6192

**Comment by Kegulf** [comment]
Ahh, that's annoying 😅 
In my opinion if one of the mentioned props is supported the renderProp should also be supported by default.

**Comment by Kegulf** [comment]
As for the render prop, you can get around it by manually passing a `data` attribute to the RAC-component, ex:

```TSX

      <ReactAriaNumberField
        name={name}
        isDisabled={disabled}
        isRequired={required}
        isReadOnly={readOnly}
        data-readOnly={readOnly} <-----
        isInvalid={isInvalid}
        minValue={min}
        maxValue={max}
        {...rest}
      >

```

**Comment by snowystinger** [comment]
Aggregating readonly issues, from https://github.com/adobe/react-spectrum/issues/9538:

### Provide a general summary of the issue here

When `isReadOnly` is set on `TextField`, the nested `<Input>` component does not receive a `data-readonly` attribute.

Input receives `data-disabled`, `data-invalid`, `data-hovered`, `data-focused` for other states, but readonly only has the native `readonly` attribute. 
This creates an inconsistent styling API where most states use `[data-*]` selectors but readonly requires the `:read-only` pseudo-class.

### 🤔 Expected Behavior?

Input should receive `data-readonly` attribute when TextField has `isReadOnly={true}`:

```html
<!-- Expected -->
<input readonly data-readonly />

CSS should work consistently:

.react-aria-Input[data-readonly] { /* should work */ }
```

### 😯 Current Behavior


Input only receives the native readonly attribute, not data-readonly:


```html
<input readonly />
```
Styling requires inconsistent approach:
```css
/* These work with data-* attributes */
.react-aria-Input[data-disabled] { }
.react-aria-Input[data-invalid] { }
.react-aria-Input[data-hovered] { }
.react-aria-Input[data-focused] { }

/* But readonly requires pseudo-class */
.react-aria-Input:read-only { }        /* workaround */
.react-aria-Input[data-readonly] { }   /* doesn't work */
```

**Comment by devongovett** [comment]
Some past PRs have improved this. I updated the table above with the current state. With #9877 we are adding support for validation props to Switch. #9892 adds isReadOnly to render props for ComboBox and NumberField. This should reach parity between render props and props.

The remaining props (Select and Slider) I don't think make much sense. Slider can't really be required because it always has a value. Validation on a slider is also strange. Readonly on Select and Slider could be supported but would also be strange and potentially confusing. If there are real use cases for those we can re-evaluate but for now with the above PRs I think we can close this.

**Comment by stefanprobst** [comment]
> for now with the above PRs I think we can close this.

i agree - maybe with the exception of readonly on Select - i have had forms which could be toggled from readonly to editable, where form elements (including selects) would inherit that shared state.

**Comment by devongovett** [comment]
A readonly select is kind of strange though. Like would the menu open when you clicked it? If not, that's kind of unexpected, and if so but you couldn't actually click anything inside it that's also weird.

**Comment by stefanprobst** [comment]
i would say the menu doesn't open, it just displays the currently selected value.

how do numberfield buttons and combobox trigger behave when readonly?

---

EDIT: although not sure what "operable" means in ["the element is not editable, but is otherwise operable"](https://w3c.github.io/aria/#aria-readonly)
